### PR TITLE
librbd: permit removal of image being bootstrapped by rbd-mirror

### PR DESCRIPTION
**Pull request by trociny**
Fixes: http://tracker.ceph.com/issues/16555
Signed-off-by: Mykola Golub <mgolub@mirantis.com>

### CODE REVIEW

File: src/librbd/internal.cc [hunk 1]
```diff
@@ -163,6 +163,46 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
   return 0;
 }
 
+void filter_out_mirror_watchers(ImageCtx *ictx,
+                                std::list<obj_watch_t> &watchers) {
```
Comment (dillaman) [inline src/librbd/internal.cc]: Nit: use a pointer instead of a reference for an in-out parameter.

File: src/test/librbd/test_mirroring.cc [hunk 2]
```diff
@@ -644,3 +644,50 @@ TEST_F(TestMirroring, MirrorStatusList) {
...
+      // received IMAGE_UPDATED notification from remove
+      if (!m_closed) {
+	FunctionContext *ctx = new FunctionContext(
+	  [this, notify_id, cookie](int r) {
+	    EXPECT_EQ(0, r);
+	    bufferlist bl;
+	    m_ictx->md_ctx.notify_ack(RBD_MIRRORING, notify_id, cookie, bl);
+	  });
+	m_closed = true;
+	m_ictx->state->close(ctx);
```
Comment (dillaman) [inline src/test/librbd/test_mirroring.cc]: Probably overkill since rbd-mirror will never be able to close the image before acking the notification.

**Comment by trociny** [comment]
@dillaman Updated.

**Comment by dillaman** [comment]
@trociny I see some random error messages popping up in the rbd-mirror daemon logs, which could be awkward if running "systemctl status rbd-mirror@XYZ" and you see lots of false errors listed. I think if this PR is combined w/ my PR for listening to the add/remove mirroring events, it might vastly reduce the chance for such false errors since the bootstrap will be canceled nearly immediately when the "DISABLING" event is received. Perhaps remove should wait for all rbd-mirror watchers to disappear before proceeding (and time-out after X seconds if they don't)? Maybe I am thinking too much into it and we can deal with that later.

**Comment by trociny** [comment]
@dillaman So, do you propose before checking for watchers (or if the check returned positive) and the mirroring is enabled to send DISABLING event and (re) check the watchers (expecting that if it was due to bootstrap then it would be cancelled)? I can do this way, rebasing it on your PR #12364 (and adding necessary bootsrap cancel bits if needed). 

On the other hand, we already send DISABLING event when disabling mirror before actually removing the image. So after the rbd-mirror is updated to cancel the bootstrap when receiving DISABLING event, I suppose the errors should gone with my current version too?

I  don't have a strong opinion what is better though.

**Comment by dillaman** [comment]
@trociny I think the chance of the race condition will be reduced (where bootstrap can complain about the image disappearing) ... and that it probably enough for the time being.

**Comment by dillaman** [comment]
retest this please

**Pull request merged**